pkg/await/apps_statefulset.go

```diff
@@ -0,0 +1,451 @@
+			reflect.TypeOf(statefulset))
+		return
+	}
+
+	// Start over, prove that rollout is complete.
+	sia.statefulsetErrors = map[string]string{}
+	sia.revisionReady = false
+	sia.replicasReady = false
+
+	// Do nothing if this is not the StatefulSet we're waiting for.
+	if statefulset.GetName() != inputStatefulSetName {
+		return
+	}
+
+	// Mark the rollout as incomplete if it's deleted.
+	if event.Type == watch.Deleted {
+		return
+	}
+
+	sia.statefulset = statefulset
+
+	// Get current generation of the StatefulSet.
+	sia.currentGeneration = statefulset.GetGeneration()
+
+	if sia.currentGeneration == 0 {
+		// No current generation, StatefulSet controller has not yet created a StatefulSet.
+		// Do nothing.
+		return
+	}
+
+	// Check if current revision matches update revision.
+	var currentRevision, updateRevision string
+	if rawCurrentRevision, ok := openapi.Pluck(statefulset.Object, "status", "currentRevision"); ok {
+		currentRevision = rawCurrentRevision.(string)
+	}
+	if rawUpdateRevision, ok := openapi.Pluck(statefulset.Object, "status", "updateRevision"); ok {
+		updateRevision = rawUpdateRevision.(string)
+	}
+	sia.revisionReady = (currentRevision != "") && (currentRevision == updateRevision)
+
+	// Check if all expected replicas are ready.
+	var replicas, statusReplicas, statusReadyReplicas, statusCurrentReplicas, statusUpdatedReplicas int64
+	if rawReplicas, ok := openapi.Pluck(statefulset.Object, "spec", "replicas"); ok {
+		replicas = rawReplicas.(int64)
+	}
+	if rawStatusReplicas, ok := openapi.Pluck(statefulset.Object, "status", "replicas"); ok {
+		statusReplicas = rawStatusReplicas.(int64)
+	}
+	if rawStatusReadyReplicas, ok := openapi.Pluck(statefulset.Object, "status", "readyReplicas"); ok {
+		statusReadyReplicas = rawStatusReadyReplicas.(int64)
+	}
+	if rawStatusCurrentReplicas, ok := openapi.Pluck(statefulset.Object, "status", "currentReplicas"); ok {
+		statusCurrentReplicas = rawStatusCurrentReplicas.(int64)
+	}
+	if rawStatusUpdatedReplicas, ok := openapi.Pluck(statefulset.Object, "status", "updatedReplicas"); ok {
+		statusUpdatedReplicas = rawStatusUpdatedReplicas.(int64)
+	} else {
+		statusUpdatedReplicas = 0
+	}
+	sia.replicasReady = (replicas == statusReplicas) &&
+		(replicas == statusReadyReplicas) &&
+		(replicas == statusCurrentReplicas)
+
+	// Set current and target replica counts for logging purposes.
+	sia.targetReplicas = replicas
+	// For the initial rollout, .status.readyReplicas is an accurate gauge of progress.
+	// For subsequent updates, we must also account for .status.updatedReplicas.
+	if sia.revisionReady {
+		sia.currentReplicas = statusReadyReplicas
+	} else {
+		// During a rollout, the number of "ready" replicas can include instances of the previous revision,
+		// so don't count those towards the target count.
+		if statusReadyReplicas > statusUpdatedReplicas {
+			sia.currentReplicas = statusUpdatedReplicas
+		} else {
+			sia.currentReplicas = statusReadyReplicas
```

Comment:
Any reason not to just always set it to `statusUpdatedReplicas`?

Reply:
`.status.updatedReplicas` gets incremented as soon as it starts to create the new Pod, but it can still fail to go ready at this point. I could change the status message to reflect that, but currently it's tracking "ready" replicas.